modules/deduplicator/src/hashtable.cpp

```diff
@@ -0,0 +1,166 @@
+/**
+ * @file
+ * @author Damir Zainullin <zaidamilda@gmail.com>
+ * @brief Definition of the HashTable class for saving flows
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include "hashtable.hpp"
+
+#include <iostream>
+#include <stdexcept>
+#include <xxhash.h>
+
+using namespace Nemea;
+
+namespace Deduplicator {
+
+HashTable::HashTable(const HashTableParameters& parameters)
+	: m_capacity(parameters.m_capacity)
+	, m_lineLength(parameters.m_lineLength)
+	, m_lineCount(parameters.m_capacity / parameters.m_lineLength)
+	, m_timeout(parameters.m_timeout)
+	, m_head(nullptr)
+	, m_tail(nullptr)
+	, m_ignored(0)
+	, m_deduplicated(0)
+	, m_storage(parameters.m_capacity)
+	, m_exportThread(&HashTable::exportExpiredThread, this)
+	, m_stop(false)
+{
+}
+
+HashTable::~HashTable() noexcept
+{
+	m_spinlock.lock();
+	m_now += std::chrono::milliseconds(m_timeout + 1);
+	m_spinlock.unlock();
+	m_stop = true;
+	m_exportThread.join();
+}
+
+bool HashTable::areDuplicateRecords(const TableRecord& tr1, const TableRecord& tr2) const noexcept
+{
+	return tr1.m_dstIp == tr2.m_dstIp && tr1.m_srcIp == tr2.m_srcIp
+		&& tr1.m_dstPort == tr2.m_dstPort && tr1.m_srcPort == tr2.m_srcPort
+		&& tr1.m_proto == tr2.m_proto && tr1.m_linkBitField != tr2.m_linkBitField
+		&& ur_timediff(tr1.m_lastTime, tr2.m_lastTime) < m_timeout;
+}
+
+void HashTable::updateUnirecIds() noexcept
+{
+	m_ids.m_srcIpId = (ur_field_id_t) ur_get_id_by_name("SRC_IP");
```

Comment:
Myslim, ze by to bylo lepsi nahradit takovouto funkci a zbavit se tak toho pretypovani a zaroven udelat kontrolu, zda je hodnota platna

```
ur_field_id_t getUnirecIDbyName(const std::string& fieldName)
{
    ur_field_id_t fieldID = ur_get_id_by_name(fieldName.c_str());
    if (fieldID == UR_E_INVALID_NAME) {
        throw std::runtime_error("...");
    }
    
    return static_cast<ur_field_id_t>(fieldID);
{
```